paasta_tools/paasta_remote_run.py

```diff
@@ -241,83 +334,210 @@ def task_config_to_dict(task_config):
+    # first, try to load credentials
     credentials_file = taskproc_config.get('boto_credential_file')
     if credentials_file:
         with open(credentials_file) as f:
             credentials = json.loads(f.read())
     else:
         raise ValueError("Required aws credentials")
 
-    if not region:
-        region = taskproc_config.get('aws_region')
-
-    endpoint = taskproc_config.get('dynamodb_endpoint')
-    session = Session(
+    # second, create the session for the given region
+    return Session(
         region_name=region,
         aws_access_key_id=credentials['accessKeyId'],
         aws_secret_access_key=credentials['secretAccessKey'],
     )
 
+
+# TODO: rename to registry?
+def build_executor_stack(
+    processor,
+    cluster_executor,
+    taskproc_config,
+    cluster,
+    region,
+):
+    """ Executor stack consists of:
+    1. Cluster Executor (e.g. MesosExecutor)
+    2. LoggingExecutor
+    3. StatefulExecutor
+    """
+    # logging executor
+    task_logging_executor = processor.executor_from_config(
+        provider='logging',
+        provider_config={
+            'downstream_executor': cluster_executor,
+        },
+    )
+    # stateful executor
     StatefulExecutor = processor.executor_cls(provider='stateful')
     stateful_executor = StatefulExecutor(
         downstream_executor=task_logging_executor,
         persister=DynamoDBPersister(
-            table_name="taskproc_events_%s" % cluster,
-            session=session,
-            endpoint_url=endpoint,
+            table_name=f"taskproc_events_{cluster}",
+            session=create_boto_session(taskproc_config, region),
+            endpoint_url=taskproc_config.get('dynamodb_endpoint'),
         ),
     )
-
     return stateful_executor
 
 
+def set_runner_signal_handlers(runner):
+    def handle_interrupt(_signum, _frame):
+        paasta_print(
+            PaastaColors.red("Signal received, shutting down scheduler."),
+        )
+        if runner is not None:
+            runner.stop()
+        sys.exit(143 if _signum == signal.SIGTERM else 1)
+    signal.signal(signal.SIGINT, handle_interrupt)
+    signal.signal(signal.SIGTERM, handle_interrupt)
+
+
+def run_task(executor, task_config):
+    """ Runs a task until a terminal event is received, which is returned. """
+    runner = Sync(executor)
+    set_runner_signal_handlers(runner)
+    terminal_event = runner.run(task_config)
+    if getattr(terminal_event, 'platform_type', None) == 'lost':
+        runner.kill(task_config.task_id)
+    runner.stop()
+    return terminal_event
+
+
+def get_terminal_event_error_message(terminal_event):
+    if not terminal_event.terminal or terminal_event.success:
+        return None  # not terminal or not an error
+    elif TASKPROC_OFFER_TIMEOUT_RAW in terminal_event.raw:
+        return terminal_event.raw
+    else:
+        mesos_type = getattr(terminal_event, 'platform_type', None)
+        if mesos_type == 'failed':
+            error_message = ''
```

Comment:
Maybe "task exited with non-zero exit code" or something, instead of leaving it empty?

Reply:
Sure.